module/zfs/spa_errlog.c

```diff
@@ -82,8 +103,91 @@ name_to_bookmark(char *buf, zbookmark_phys_t *zb)
 	zb->zb_blkid = zfs_strtonum(buf + 1, &buf);
 	ASSERT(*buf == '\0');
 }
+
+/*
+ * Convert a string to a err_phys.
+ */
+static void
+name_to_errphys(char *buf, zbookmark_err_phys_t *zep)
+{
+	zep->zb_object = zfs_strtonum(buf, &buf);
+	ASSERT(*buf == ':');
+	zep->zb_level = (int)zfs_strtonum(buf + 1, &buf);
+	ASSERT(*buf == ':');
+	zep->zb_blkid = zfs_strtonum(buf + 1, &buf);
+	ASSERT(*buf == ':');
+	zep->zb_birth = zfs_strtonum(buf + 1, &buf);
+	ASSERT(*buf == '\0');
+}
+
+static void
+zeb_to_zb(uint64_t dataset, zbookmark_err_phys_t *zep, zbookmark_phys_t *zb)
+{
+	zb->zb_objset = dataset;
+	zb->zb_object = zep->zb_object;
+	zb->zb_level = zep->zb_level;
+	zb->zb_blkid = zep->zb_blkid;
+}
 #endif
 
+static void
+name_to_object(char *buf, uint64_t *obj)
+{
+	*obj = zfs_strtonum(buf, &buf);
+	ASSERT(*buf == '\0');
+}
+
+static int
+get_head_and_birth_txg(spa_t *spa, zbookmark_err_phys_t *zep, uint64_t ds_obj,
+    uint64_t *head_dataset_id)
+{
+	dsl_pool_t *dp = spa->spa_dsl_pool;
+	dsl_dataset_t *ds;
+	objset_t *os;
+
+	dsl_pool_config_enter(dp, FTAG);
+	int err = dsl_dataset_hold_obj(dp, ds_obj, FTAG, &ds);
+	if (err != 0) {
+		dsl_pool_config_exit(dp, FTAG);
+		return (EFAULT);
+	}
+	*head_dataset_id = dsl_dir_phys(ds->ds_dir)->dd_head_dataset_obj;
```

Comment:
should we assert that head_dataset_id is not null, before dereferencing it?

Reply:
It doesn't show here in the conversation, but I added the assertion.